### PR TITLE
Save installation domain when creating or connecting Nosto account

### LINKED ISSUE #382: Save installation domain when creating or connecting Nosto account

### PR DESCRIPTION
**Pull request by olsiqose-nosto**
## Description
Store the storeview domain in database when installing or connecting Nosto account.
Display system notification if the domain from the database does not match the real storeview domain. 

## Related Issue
closes 
#382
#384

## Motivation and Context
Sometimes merchant create staging environment by copying the production one. This leads to having the same Nosto account for both environments. 

## How Has This Been Tested?
Tested manually by creating account. 
Changed domain and checked if notification was working properly.

## Checklist:

- [x] My code follows the code style of this project.
- [x] I have assigned the correct milestone or created one if non existent.
- [x] I have correctly labeled this pull request.
- [x] I have linked the corresponding issue in this description.
- [x] I have updated the corresponding Jira ticket.
- [x] I have requested a review from at least 2 reviewers

### CODE REVIEW

File: Model/Admin/Notification/Messages.php [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+     */
+    public function isDisplayed()
+    {
+        $stores = $this->nostoHelperScope->getStores();
+
+        foreach ($stores as $store) {
+
+            //Check if the store is connected to Nosto
+            if ($this->nostoHelperAccount->findAccount($store)) {
+                if ($this->nostoHelperAccount->isDomainValid($store) === false) {
```
Comment (supercid) [inline Model/Admin/Notification/Messages.php]: Merge if's to avoid nesting

File: Helper/Account.php [hunk 2]
```diff
@@ -86,14 +93,16 @@ class Account extends AbstractHelper
     public function __construct(
         Context $context,
         WriterInterface $appConfig,
-        NostoHelperScope $nostoHelperScope
+        NostoHelperScope $nostoHelperScope,
+        NostoHelperUrl $nostoHelperUrl
```
Comment (supercid) [inline Helper/Account.php]: Missing phpdoc

File: Helper/Account.php [hunk 3]
```diff
@@ -283,4 +303,17 @@ public function getStoresWithNosto()
...
+     * Returns bool value that represent validity of domain
+     *
+     * @param Store $store
+     * @return bool
+     */
+    public function isDomainValid(Store $store)
+    {
+        $storedDomain = $store->getConfig(self::XML_PATH_DOMAIN);
+        $realDomain = $this->nostoHelperUrl->getActiveDomain($store);
+        return ($realDomain == $storedDomain);
```
Comment (supercid) [inline Helper/Account.php]: Strict types, use `===`

File: Model/Admin/Notification/Messages.php [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+        $this->nostoHelperScope = $nostoHelperScope;
+        $this->nostoHelperAccount = $nostoHelperAccount;
+    }
+
+    /**
+     * @return \Magento\Framework\Phrase|string
+     */
+    public function getText()
+    {
+        $message = __($this->message);
```
Comment (supercid) [inline Model/Admin/Notification/Messages.php]: Redundant variable, you can return directly

File: Model/Admin/Notification/Messages.php [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+    }
+
+    /**
+     * Set the value of the message
+     */
+    private function buildMessage(){
+        $message = 'Nosto account is invalid for the stores: ';
+
+        foreach ($this->storeNames as $storeName) {
+            $message .= " * ".$storeName;
```
Comment (supercid) [inline Model/Admin/Notification/Messages.php]: Single quotes, since you're not evaluating any expression

File: etc/adminhtml/di.xml [hunk 6]
```diff
@@ -0,0 +1,11 @@
+<?xml version="1.0" encoding="utf-8"?>
```
Comment (supercid) [inline etc/adminhtml/di.xml]: Missing copyright headers

File: Model/Admin/Notification/Messages.php [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+    public function __construct(
+        NostoHelperAccount $nostoHelperAccount,
+        NostoHelperScope $nostoHelperScope
+    ) {
+        $this->nostoHelperScope = $nostoHelperScope;
+        $this->nostoHelperAccount = $nostoHelperAccount;
+    }
+
+    /**
+     * @return \Magento\Framework\Phrase|string
```
Comment (hannunosto) [inline Model/Admin/Notification/Messages.php]: Import this one

File: Model/Admin/Notification/Messages.php [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+     */
+    public function isDisplayed()
+    {
+        $stores = $this->nostoHelperScope->getStores();
+
+        foreach ($stores as $store) {
+            //Check if the store is connected to Nosto
+            if ($this->nostoHelperAccount->findAccount($store)
+                && $this->nostoHelperAccount->isDomainValid($store) === false) {
+                    $this->storeNames[] = $store->getName();
```
Comment (hannunosto) [inline Model/Admin/Notification/Messages.php]: Is there a reason why `storeNames` is a class variable? Could it just be a local variable that you'd pass to `buildMessage`?

File: Model/Admin/Notification/Messages.php [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+        // From here you can change notification message type.
+        return MessageInterface::SEVERITY_CRITICAL;
+    }
+
+    /**
+     * Set the value of the message
+     */
+    private function buildMessage()
+    {
+        $message = 'Nosto account is invalid for the stores: ';
```
Comment (hannunosto) [inline Model/Admin/Notification/Messages.php]: I think we should run these through the translator

File: Model/Admin/Notification/Messages.php [hunk 10]
```diff
@@ -0,0 +1,127 @@
...
+
+        foreach ($this->storeNames as $storeName) {
+            $message .= ' * ' .$storeName;
+        }
+
+        $message .= '. Please re-login';
```
Comment (hannunosto) [inline Model/Admin/Notification/Messages.php]: What does this mean? Login to the store admin again?

File: Model/Admin/Notification/Messages.php [hunk 11]
```diff
@@ -0,0 +1,127 @@
...
+     */
+    public function isDisplayed()
+    {
+        $stores = $this->nostoHelperScope->getStores();
+        $storeNames = [];
+
+        foreach ($stores as $store) {
+            //Check if the store is connected to Nosto
+            if ($this->nostoHelperAccount->findAccount($store)
+                && $this->nostoHelperAccount->isDomainValid($store) === false) {
```
Comment (supercid) [inline Model/Admin/Notification/Messages.php]: This is unindented. Closing brace and opening curly brace `) {` should go on the next line and block inside should be back 1 tab

File: Model/Admin/Notification/Messages.php [hunk 12]
```diff
@@ -0,0 +1,127 @@
...
+
+    /**
+     * Set the value of the message
+     */
+    private function buildMessage($storeNames)
+    {
+        $message = 'Nosto account is invalid for the stores: ';
+
+        foreach ($storeNames as $storeName) {
+            $message .= ' * ' .$storeName;
```
Comment (supercid) [inline Model/Admin/Notification/Messages.php]: Missing space after `.`

File: Model/Admin/Notification/Messages.php [hunk 13]
```diff
@@ -0,0 +1,127 @@
...
+     */
+    public function getSeverity()
+    {
+        return MessageInterface::SEVERITY_CRITICAL;
+    }
+
+    /**
+     * Set the value of the message
+     */
+    private function buildMessage($storeNames)
```
Comment (supercid) [inline Model/Admin/Notification/Messages.php]: Missing phpdoc for `$storeNames`

File: Model/Admin/Notification/Messages.php (hunk 9 above)
Comment (mridang) [reply Model/Admin/Notification/Messages.php]: 👍

**Pull request merged**